Support digest auth

shankardevy opened the pull request:
@marcoherbst I have a PR to HTTPotion with successful test for httpdigest
see https://github.com/myfreeweb/httpotion/pull/61

I am using the same code here with some additions needed because of evercam-media's current limitation of not knowing the needed auth type for the given camera.

I have tested the http_digest with httpotion, so I assume it will also work here. @mosic can you please test this and let me know if you get any error?

marcoherbst commented:
Sounds good, well done. @mosic will reply.

M

On 1 July 2015 at 06:52, Shankar Dhanasekaran <notifications@github.com>
wrote:

> @marcoherbst <https://github.com/marcoherbst> I have a PR to HTTPotion
> with successful test for httpdigest
> see myfreeweb/httpotion#61
> <https://github.com/myfreeweb/httpotion/pull/61>
>
> I am using the same code here with some additions needed because of
> evercam-media's current limitation of not knowing the needed auth type for
> the given camera.
>
> I have tested the http_digest with httpotion, so I assume it will also
> work here. @mosic <https://github.com/mosic> can you please test this and
> let me know if you get any error?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/evercam/evercam-media/pull/12
> Commit Summary
>
>    - Support digest auth
>
> File Changes
>
>    - *A* lib/evercam_media/http_client.ex
>    <https://github.com/evercam/evercam-media/pull/12/files#diff-0> (115)
>    - *M* lib/snapshots/snapshot_fetch.ex
>    <https://github.com/evercam/evercam-media/pull/12/files#diff-1> (15)
>    - *M* web/controllers/snapshot_controller.ex
>    <https://github.com/evercam/evercam-media/pull/12/files#diff-2> (3)
>
> Patch Links:
>
>    - https://github.com/evercam/evercam-media/pull/12.patch
>    - https://github.com/evercam/evercam-media/pull/12.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/pull/12>.
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst

shankardevy commented:
@mosic are you suggesting this? https://github.com/shankardevy/evercam-media/commit/94ce60299cef412e06a13e7e2c9985eaafae0660

mosic commented:
@shankardevy I tested the pull request and it seems to be working fine for most of the cameras but it does error on a few. One of them is `evercam-remembrance-camera`, demo camera we share with newly registered users. Here is the error I got:

```
[error] %MatchError{term: %{"realm" => "API"}}
[error]     (evercam_media) lib/evercam_media/http_client.ex:52: EvercamMedia.HTTPClient.DigestAuth.get_digest_token/4
    (evercam_media) lib/evercam_media/http_client.ex:29: EvercamMedia.HTTPClient.get/5
    (evercam_media) web/controllers/snapshot_controller.ex:75: EvercamMedia.SnapshotController.snapshot/3
    (evercam_media) web/controllers/snapshot_controller.ex:12: EvercamMedia.SnapshotController.show/2
    (evercam_media) web/controllers/snapshot_controller.ex:1: EvercamMedia.SnapshotController.phoenix_controller_pipeline/2
    (evercam_media) lib/phoenix/router.ex:297: EvercamMedia.Router.dispatch/2
    (evercam_media) lib/phoenix/router.ex:2: EvercamMedia.Router.call/2
    (evercam_media) lib/media/endpoint.ex:1: EvercamMedia.Endpoint.phoenix_endpoint_pipeline/1
```

marcoherbst commented:
That's odd, as this is not a digest camera.

this is a standard hikvision.

@mosic are you sure this error was not happening already? before the code
change?

The only difference with this cam is that perhaps it is getting slightly
more requests than most.

Also, beware that a hikvision camera can have 2 alternate urls ... the one
we like is streaming/channel/1/picture , the one we dont like (slower) is
/onvif/snapshot

M



On 2 July 2015 at 02:00, Miloš Mošić <notifications@github.com> wrote:

> @shankardevy <https://github.com/shankardevy> I tested the pull request
> and it seems to be working fine for most of the cameras but it does error
> on a few. One of them is evercam-remembrance-camera, demo camera we share
> with newly registered users. Here is the error I got:
>
> [error] %MatchError{term: %{"realm" => "API"}}
> [error]     (evercam_media) lib/evercam_media/http_client.ex:52: EvercamMedia.HTTPClient.DigestAuth.get_digest_token/4
>     (evercam_media) lib/evercam_media/http_client.ex:29: EvercamMedia.HTTPClient.get/5
>     (evercam_media) web/controllers/snapshot_controller.ex:75: EvercamMedia.SnapshotController.snapshot/3
>     (evercam_media) web/controllers/snapshot_controller.ex:12: EvercamMedia.SnapshotController.show/2
>     (evercam_media) web/controllers/snapshot_controller.ex:1: EvercamMedia.SnapshotController.phoenix_controller_pipeline/2
>     (evercam_media) lib/phoenix/router.ex:297: EvercamMedia.Router.dispatch/2
>     (evercam_media) lib/phoenix/router.ex:2: EvercamMedia.Router.call/2
>     (evercam_media) lib/media/endpoint.ex:1: EvercamMedia.Endpoint.phoenix_endpoint_pipeline/1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/pull/12#issuecomment-117865930>.
>



-- 
Marco Herbst
*Evercam.io <http://www.evercam.io/>* - Cameras Unleashed
Mobile: +353 86 196 0451
Skype: marco.a.herbst

mosic commented:
@marcoherbst Yes, I'm sure. I agree it's odd though.

shankardevy commented:
Thanks @mosic for the review. I shall check and update the PR shortly.

On Thu, Jul 2, 2015 at 6:30 AM, Miloš Mošić <notifications@github.com>
wrote:

> @shankardevy <https://github.com/shankardevy> I tested the pull request
> and it seems to be working fine for most of the cameras but it does error
> on a few. One of them is evercam-remembrance-camera, demo camera we share
> with newly registered users. Here is the error I got:
>
> [error] %MatchError{term: %{"realm" => "API"}}
> [error]     (evercam_media) lib/evercam_media/http_client.ex:52: EvercamMedia.HTTPClient.DigestAuth.get_digest_token/4
>     (evercam_media) lib/evercam_media/http_client.ex:29: EvercamMedia.HTTPClient.get/5
>     (evercam_media) web/controllers/snapshot_controller.ex:75: EvercamMedia.SnapshotController.snapshot/3
>     (evercam_media) web/controllers/snapshot_controller.ex:12: EvercamMedia.SnapshotController.show/2
>     (evercam_media) web/controllers/snapshot_controller.ex:1: EvercamMedia.SnapshotController.phoenix_controller_pipeline/2
>     (evercam_media) lib/phoenix/router.ex:297: EvercamMedia.Router.dispatch/2
>     (evercam_media) lib/phoenix/router.ex:2: EvercamMedia.Router.call/2
>     (evercam_media) lib/media/endpoint.ex:1: EvercamMedia.Endpoint.phoenix_endpoint_pipeline/1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/pull/12#issuecomment-117865930>.
>

shankardevy commented:
@mosic with the above commit, it should not pick up the right authentication for the camera. The error you have reported stems from the fact that the previous code tries a digest auth when a basic auth fails due to either incorrect password or url. can you please check with this commit for all types of camera and let me know if you still get any error?

If you are getting any error still, could you please let me know the full url to the snapshot, auth type, credentials by email?

jgarciasacristan commented:
@shankardevy  I have made some modifications to your httpotion  changes to support post requests. You can see my changes [here] (https://github.com/jgarciasacristan/httpotion/commit/4e1fb929ee4ce283e13fea5088c06ee1071184ff) 
I am working on a [project] (https://github.com/jgarciasacristan/onvifex) to support OnVIF commands and support for digest authentication is required. Only digest authentication or WS-Security are supported in OnVIF (See page 30 of [OnVIF Core Spec] (http://www.onvif.org/specs/core/ONVIF-Core-Specification-v221.pdf) )
The camera that I am using in my tests requires digest authentication. Unfortunately after my changes to httpotion still does not work. 
I would really appreciate if you can download my project and give it a try whenever you have time. It may be an easy fix but it will make a big difference to me. Please be aware that GetDateAndTime operation works so the SOAP commands are correct as far as I know.
In the meantime I will work in the support for WS-Security.
Thank you

jgarciasacristan commented:
@shankardevy I have managed to talk OnViF with the hikvision camera using WS-Security headers. No need of digest authentication from my part as I am using WS-Security. Please ignore previous comment.

shankardevy commented:
@jgarciasacristan thanks for the update. Sorry, I was engaged with  many
things during the time you asked for my help. I am available these days to
work on cams/elixir.

On Mon, Jul 20, 2015 at 4:34 PM, jgarciasacristan <notifications@github.com>
wrote:

> @shankardevy <https://github.com/shankardevy> I have managed to talk
> OnViF with the hikvision camera using WS-Security headers. No need of
> digest authentication from my part as I am using WS-Security. Please ignore
> previous comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/pull/12#issuecomment-122849956>.
>

jgarciasacristan commented:
Thank you Shankar,
No worries. I am happy that WSSE works. I won't hesitate to contact you
when I have issues. :-)
Regards
Jose

On Mon, Jul 20, 2015 at 5:56 PM, Shankar Dhanasekaran <
notifications@github.com> wrote:

> @jgarciasacristan thanks for the update. Sorry, I was engaged with many
> things during the time you asked for my help. I am available these days to
> work on cams/elixir.
>
> On Mon, Jul 20, 2015 at 4:34 PM, jgarciasacristan <
> notifications@github.com>
> wrote:
>
> > @shankardevy <https://github.com/shankardevy> I have managed to talk
> > OnViF with the hikvision camera using WS-Security headers. No need of
> > digest authentication from my part as I am using WS-Security. Please
> ignore
> > previous comment.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/evercam/evercam-media/pull/12#issuecomment-122849956
> >.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/evercam/evercam-media/pull/12#issuecomment-122950022>.
>

mosic commented on web/controllers/snapshot_controller.ex:
```diff
@@ -62,8 +63,18 @@ defmodule EvercamMedia.SnapshotController do
-      data = fetch(url, auth)
+      [username, password] = String.split(auth, ":")
+      camera = Camera.by_exid(camera_id) |> EvercamMedia.Repo.one |> EvercamMedia.Repo.preload(:vendor_model)
+      vendor_model = camera.vendor_model |> EvercamMedia.Repo.preload(:vendor)
```
This can fail when a camera doesn't have a `vendor_model`, it would return a `FunctionClauseError`:

```
[error] %FunctionClauseError{arity: 3, function: :preload, module: Ecto.Repo.Preloader}
[error]     (ecto) lib/ecto/repo/preloader.ex:35: Ecto.Repo.Preloader.preload(nil, EvercamMedia.Repo, :vendor)
```

mosic commented on lib/snapshots/snapshot_fetch.ex:
```diff
@@ -16,7 +16,7 @@ defmodule EvercamMedia.Snapshot do
       request.body
     end
-  end
+  end		
```
Remove this whitespace.

mosic commented:
@shankardevy This looks good aside from the two comments I just left.

I see that you haven't modified `check_camera` function used in the background workers to use the same method of choosing auth types. Were you planning on making a separate pull request for that?

shankardevy commented:
@mosic I have modified the query to avoid too many conditionals checks in the controller and also moved the query to model. The query should return only vendor's exid or nil. This should resolve the issue faced and will fall back to basic auth then when nil is returned.

shankardevy commented:
@mosic  
> I see that you haven't modified check_camera function used in the background workers.

I see an issue related to modifying how we check a camera. should this wait until we decide on it or shall I provide a separate another PR for it now?

mosic commented:
@shankardevy That's great, thanks. You can update `check_camera` to use the same code for now and we will discuss later how the process of checking a camera can be improved.
